### DISCUSSION TITLE
Still a lot of re-renders in v9

### DISCUSSION THREAD

**Discussion by ArthurGoupil** [gpbl/react-day-picker#2255] [General]
## Description

Hello!

I was following this issue for quite a long time https://github.com/gpbl/react-day-picker/issues/1473 and was happy to see it has been potentially fixed in v9.

From my first tests, it looks like when selecting a date, it still re-renders every days of every months:

https://github.com/gpbl/react-day-picker/assets/50388474/d7d0b109-834a-4e67-8c1b-88f7dd5ef129

Here is the [related sandbox](https://codesandbox.io/p/sandbox/rendering-test-react-day-picker-v9-3kwpm8?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxswoprs00063b85f1xkepb7%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxswoprs00023b853iraud9f%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxswoprs00033b853vu702mg%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxswoprs00053b85m3x4szt8%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxswoprs00023b853iraud9f%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxswoprs00013b85v5jnbdo3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxswoprs00023b853iraud9f%2522%252C%2522activeTabId%2522%253A%2522clxswoprs00013b85v5jnbdo3%2522%257D%252C%2522clxswoprs00053b85m3x4szt8%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxswoprs00043b854osz6xuh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clxswoprs00053b85m3x4szt8%2522%252C%2522activeTabId%2522%253A%2522clxswoprs00043b854osz6xuh%2522%257D%252C%2522clxswoprs00033b853vu702mg%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clxswoprs00033b853vu702mg%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) 

Am I missing something in the basic implementation or so?

Thanks a lot!

## Expected Behavior

It does not re-render every days when not necessary.

**Comment by gpbl** [comment]
Hey @ArthurGoupil, thanks for the heads up and for testing v9.

The rerender is necessary to manage focus: when a day is clicked, the other cells need to know if they should become the "possible focus target."

We may explore other ways to handle focus, but they are challenging to implement following React patterns.

I believe the rerendering shouldn't be too concerning. I've moved out some of the calculations that made rerendering slow in the previous version. There are some upcoming changes that will further memoize the values used by DayPicker, thus reducing other unnecessary rerenders.

**Comment by ArthurGoupil** [comment]
Alright thanks a lot for the answer @gpbl 🙏

**Comment by gpbl** [comment]
It is true there are still some useless re-renders when interacting with the calendar. For sure some of those are avoidable using `React.memo` or similar utilities. We need to setup an environment for testing the performance first: these tools may affect usage of memory too and most of the time they are not worth it.

If you have some suggestion give a look at https://github.com/gpbl/react-day-picker/blob/main/src/DayPicker.tsx. Modifiers, focus, creating new objects via spread are likely the reason the re-renders happen. We may also want to debounce keyboard interaction...

I don't think these rerendering should be a concern for the release, but I can't exclude some improvements before the release. They can come as patch or minor upgrades once we have enough feedback from the users.

Thanks for the heads-up - I'm moving this to the discussions.
